File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.py
class MedianFinder:

    def __init__(self):
        self.minHeap = []
        self.maxHeap = []
        self.even = True
        

    def addNum(self, num: int) -> None:
        if(self.even):
            p = -heapq.heappushpop(self.minHeap, num)
            heapq.heappush(self.maxHeap, p)
        else:
            p = -heapq.heappushpop(self.maxHeap, -num)
            heapq.heappush(self.minHeap, p)   
        self.even = not self.even
    def findMedian(self) -> float:
        if not self.even:
            return -self.maxHeap[0]
        else:
            return (-self.maxHeap[0]+self.minHeap[0])/2
        


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.java
class MedianFinder {
    PriorityQueue<Integer> minH;
    PriorityQueue<Integer> maxH;
    boolean isEven;
    public MedianFinder() {
        minH = new PriorityQueue<>();
        maxH = new PriorityQueue<>((a, b)->b-a);
        isEven = true;
    }
    
    public void addNum(int num) {
        if(isEven){
            minH.add(num);
            maxH.add(minH.poll());
            
        }
        else{
            maxH.add(num);
            minH.add(maxH.poll());
        }
        isEven = !isEven;
        
    }
    
    public double findMedian() {
        if(isEven){
            return (minH.peek()+maxH.peek())/2.0;
        }
        else{
            return (double)maxH.peek();
        }
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */